be/src/agent/publish_version.cpp

```diff
@@ -0,0 +1,152 @@
+        num_tablet += partitions[i].size();
+    }
+    std::vector<TabletPublishVersionTask> tablet_tasks(num_tablet);
+    size_t tablet_idx = 0;
+    for (size_t i = 0; i < publish_version_req.partition_version_infos.size(); i++) {
+        for (auto& itr : partitions[i]) {
+            auto& task = tablet_tasks[tablet_idx++];
+            task.txn_id = transaction_id;
+            task.partition_id = publish_version_req.partition_version_infos[i].partition_id;
+            task.tablet_id = itr.first.tablet_id;
+            task.version = publish_version_req.partition_version_infos[i].version;
+            task.rowset = std::move(itr.second);
+        }
+    }
+    std::mutex affected_dirs_lock;
+    for (size_t i = 0; i < tablet_tasks.size(); i++) {
+        uint32_t retry_time = 0;
+        Status st;
+        while (retry_time++ < PUBLISH_VERSION_SUBMIT_MAX_RETRY) {
+            st = threadpool.submit_func([&, i]() {
+                auto& task = tablet_tasks[i];
+                if (!task.rowset) {
+                    task.st = Status::NotFound(
+                            fmt::format("rowset not found  of tablet: {}, txn_id: {}", task.tablet_id, task.txn_id));
+                    LOG(WARNING) << task.st;
+                    return;
+                }
+                TabletSharedPtr tablet = StorageEngine::instance()->tablet_manager()->get_tablet(task.tablet_id);
+                if (!tablet) {
+                    task.st = Status::NotFound(
+                            fmt::format("Not found tablet to publish_version. tablet_id: {}, txn_id: {}",
+                                        task.tablet_id, task.txn_id));
+                    LOG(WARNING) << task.st;
+                    return;
+                }
+                {
+                    std::lock_guard lg(affected_dirs_lock);
+                    affected_dirs.insert(tablet->data_dir());
+                }
+                task.st = StorageEngine::instance()->txn_manager()->publish_txn(task.partition_id, tablet, task.txn_id,
+                                                                                task.version, task.rowset);
+                if (!task.st.ok()) {
+                    LOG(WARNING) << "Publish txn failed tablet:" << tablet->tablet_id() << " version:" << task.version
+                                 << " partition:" << task.partition_id << " txn_id: " << task.txn_id
+                                 << " rowset:" << task.rowset->rowset_id();
+                } else {
+                    LOG(INFO) << "Publish txn success tablet:" << tablet->tablet_id() << " version:" << task.version
+                              << " partition:" << task.partition_id << " txn_id: " << task.txn_id
+                              << " rowset:" << task.rowset->rowset_id();
+                }
+                task.version = tablet->max_continuous_version_from_beginning().second;
```

Comment:
this operation is quite heavy, it don't think we should do it in publish.

Reply:
updated, use an optimized version `max_continuous_version`